<launch>

  <arg name="sensorOffsetX" default="0"/>
  <arg name="sensorOffsetY" default="0"/>
  <arg name="cameraOffsetZ" default="0"/>
  <arg name="twoWayDrive" default="false"/>
  <arg name="maxSpeed" default="2.0"/>
  <arg name="autonomyMode" default="true"/>
  <arg name="autonomySpeed" default="2.0"/>
  <arg name="joyToSpeedDelay" default="2.0"/>
  <arg name="goalX" default="0"/>
  <arg name="goalY" default="0"/>

  <node pkg="local_planner" type="localPlanner" name="localPlanner" output="log" required="true">
    
    <remap from="/state_estimation" to="/laser_odom_init"/>
    <remap from="/registered_scan" to="/registered_point_cloud"/>
    <remap from="/way_point" to="/path_execution_node/look_ahead_goal"/>

    <param name="pathFolder" type="string" value="$(find local_planner)/paths" />
    <param name="vehicleLength" type="double" value="0.4" />
    <param name="vehicleWidth" type="double" value="0.3" />
    <param name="sensorOffsetX" value="$(arg sensorOffsetX)" />
    <param name="sensorOffsetY" value="$(arg sensorOffsetY)" />
    <param name="twoWayDrive" value="$(arg twoWayDrive)" />
    <param name="laserVoxelSize" type="double" value="0.05" />
    <param name="terrainVoxelSize" type="double" value="0.2" />
    <param name="useTerrainAnalysis" type="bool" value="true" />
    <param name="checkObstacle" type="bool" value="true" />
    <param name="checkRotObstacle" type="bool" value="false" />
    <param name="adjacentRange" type="double" value="4.25" />
    <param name="obstacleHeightThre" type="double" value="0.15" />
    <param name="groundHeightThre" type="double" value="0.1" />
    <param name="costHeightThre" type="double" value="0.1" />
    <param name="costScore" type="double" value="0.02" />
    <param name="useCost" type="bool" value="false" />
    <param name="pointPerPathThre" type="int" value="2" />
    <param name="minRelZ" type="double" value="-0.52" />
    <param name="maxRelZ" type="double" value="0.25" />
    <param name="maxSpeed" value="$(arg maxSpeed)" />
    <param name="dirWeight" type="double" value="0.02" />
    <param name="dirThre" type="double" value="90.0" />
    <param name="dirToVehicle" type="bool" value="false" />
    <param name="pathScale" type="double" value="1.25" />
    <param name="minPathScale" type="double" value="0.75" />
    <param name="pathScaleStep" type="double" value="0.25" />
    <param name="pathScaleBySpeed" type="bool" value="true" />
    <param name="minPathRange" type="double" value="1.0" />
    <param name="pathRangeStep" type="double" value="0.5" />
    <param name="pathRangeBySpeed" type="bool" value="true" />
    <param name="pathCropByGoal" type="bool" value="true" />
    <param name="autonomyMode" value="$(arg autonomyMode)" />
    <param name="autonomySpeed" value="$(arg autonomySpeed)" />
    <param name="joyToSpeedDelay" value="$(arg joyToSpeedDelay)" />
    <param name="joyToCheckObstacleDelay" type="double" value="5.0" />
    <param name="goalClearRange" type="double" value="0.5" />
    <param name="goalX" type="double" value="$(arg goalX)" />
    <param name="goalY" type="double" value="$(arg goalY)" />
  </node>

  <node pkg="local_planner" type="pathFollower" name="pathFollower" output="log" required="true">

    <remap from="/state_estimation" to="/laser_odom_init"/>

    <param name="sensorOffsetX" value="$(arg sensorOffsetX)" />
    <param name="sensorOffsetY" value="$(arg sensorOffsetY)" />
    <param name="pubSkipNum" type="int" value="1" />
    <param name="twoWayDrive" value="$(arg twoWayDrive)" />
    <param name="lookAheadDis" type="double" value="0.5" />
    <param name="yawRateGain" type="double" value="7.5" />
    <param name="stopYawRateGain" type="double" value="7.5" />
    <param name="maxYawRate" type="double" value="60.0" />
    <param name="maxYawAccel" type="double" value="10.0" />
    <param name="maxSpeed" value="$(arg maxSpeed)" />
    <param name="maxAccel" type="double" value="2.5" />
    <param name="switchTimeThre" type="double" value="1.0" />
    <param name="dirDiffThre" type="double" value="0.1" />
    <param name="stopDisThre" type="double" value="0.2" />
    <param name="slowDwnDisThre" type="double" value="1.0" />
    <param name="useInclRateToSlow" type="bool" value="false" />
    <param name="inclRateThre" type="double" value="120.0" />
    <param name="slowRate1" type="double" value="0.25" />
    <param name="slowRate2" type="double" value="0.5" />
    <param name="slowTime1" type="double" value="2.0" />
    <param name="slowTime2" type="double" value="2.0" />
    <param name="useInclToStop" type="bool" value="false" />
    <param name="inclThre" type="double" value="45.0" />
    <param name="stopTime" type="double" value="5.0" />
    <param name="noRotAtStop" type="bool" value="false" />
    <param name="noRotAtGoal" type="bool" value="true" />
    <param name="autonomyMode" value="$(arg autonomyMode)" />
    <param name="autonomySpeed" value="$(arg autonomySpeed)" />
    <param name="joyToSpeedDelay" value="$(arg joyToSpeedDelay)" />
  </node>

</launch>
